bindings/java/src/main/com/apple/foundationdb/MappedRangeQueryV2.java

```diff
@@ -0,0 +1,353 @@
+/*
+ * RangeQuery.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb;
+
+import com.apple.foundationdb.EventKeeper.Events;
+import com.apple.foundationdb.async.AsyncIterable;
+import com.apple.foundationdb.async.AsyncIterator;
+import com.apple.foundationdb.async.AsyncUtil;
+
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+
+// TODO: Share code with RangeQuery?
+/**
+ * Represents a query against FoundationDB for a range of keys. The
+ *  result of this query can be iterated over in a blocking fashion with a call to
+ *  {@link #iterator()} (as specified by {@link Iterable}).
+ *  If the calling program uses an asynchronous paradigm, a non-blocking
+ *  {@link AsyncIterator} is returned from {@link #iterator()}. Both of these
+ *  constructions will not begin to query the database until the first call to
+ *  {@code hasNext()}. As the query uses its {@link Transaction} of origin to fetch
+ *  all the data, the use of this query object must not span more than a few seconds.
+ *
+ * <br><br><b>NOTE:</b> although resulting {@code Iterator}s do support the {@code remove()}
+ *   operation, the remove is not durable until {@code commit()} on the {@code Transaction}
+ *   that yielded this query returns <code>true</code>.
+ */
+class MappedRangeQueryV2 implements AsyncIterable<MappedKeyValueV2> {
```

Comment:
There's a lot of duplication between this class and `MappedRangeQuery`. (That's true for a lot of these classes introduced with a `V2` suffix.) I think ideally, they would be collapsed into just one class (by having the Java bindings or C bindings handle the differences between the v1 and v2 APIs at a lower level), but if they're not, it seems like this should be structured in a way to allow for more code-reuse. For example, there could be an `AbstractMappedRangeQuery<K>` which concrete implementations `MappedRangeQuery extends AbstractMappedRangeQuery<MappedKeyValue>` and `MappedRangeQueryV2 extends AbstractMappedRangeQuery<MappedKeyValueV2>`. And similarly for other classes

Reply:
I do not think 1 class can do the job, `MappedKeyValueRefV2` and `MappedKeyValueRef`  have to be 2 classes because one of them has an additional field (a bytes array), then C also has to have 2 versions, etc.

to restructure it is a good idea, I will work on it once we resolve other more critical issues.

Reply:
Hm, what if `MappedKeyValueRef` was changed to always have the new field, and then in the v1 API function, it moved the data out of a `MappedKeyValueRef` into something like a `MappedKeyValueRefV1` that didn't have the new field? That way, anyone calling the v1 method gets data back in the format they expect, but there don't need to be duplicate implementations throughout the code?

I also think that even if you didn't do that, you could still have either a templated code that works with either a `MappedKeyValueRef` or a `MappedKeyValueRefV2`, or `MappedKeyValueRefV2` could inherit from `MappedKeyValueRef`.